File: web/src/api/FeatureApi.ts
import Api, { apiState } from 'api/Api';
import { selector } from 'recoil';
import FeatureRep from 'rep/FeatureRep';

class FeatureApi {
  private readonly api: Api;

  constructor(api: Api) {
    this.api = api;
  }

  async listByOrganization(organizationGuid: string): Promise<FeatureRep[]> {
    return await this.api.request<FeatureRep[]>({
      method: 'GET',
      path: `/organizations/${organizationGuid}/features`,
    });
  }
}

export const featureApiState = selector<FeatureApi>({
  key: 'api/feature',
  get: ({ get }) => {
    const api = get(apiState({ authenticated: true }));
    return new FeatureApi(api);
  },
});

File: feature/organization/interface/src/main/kotlin/limber/api/feature/FeatureApi.kt
package limber.api.feature

import io.ktor.http.HttpMethod
import jakarta.validation.Valid
import limber.feature.rest.RestEndpoint
import limber.rep.feature.FeatureRep
import java.util.UUID

public object FeatureApi {
  public data class Get(
    val featureId: String,
  ) : RestEndpoint<Nothing>() {
    override val method: HttpMethod = HttpMethod.Get
    override val path: String = "/features/$featureId"
  }

  public data class ListByOrganization(
    val organizationGuid: UUID,
  ) : RestEndpoint<Nothing>() {
    override val method: HttpMethod = HttpMethod.Get
    override val path: String = "/organizations/$organizationGuid/features"
  }

  public data class Create(
    val organizationGuid: UUID,
    @Valid override val body: FeatureRep.Creator?,
  ) : RestEndpoint<FeatureRep.Creator>() {
    override val method: HttpMethod = HttpMethod.Post
    override val path: String = "/organizations/$organizationGuid/features"
  }

  public data class Update(
    val featureId: String,
    @Valid override val body: FeatureRep.Update?,
  ) : RestEndpoint<FeatureRep.Update>() {
    override val method: HttpMethod = HttpMethod.Patch
    override val path: String = "/features/$featureId"
  }

  public data class Delete(
    val featureId: String,
  ) : RestEndpoint<Nothing>() {
    override val method: HttpMethod = HttpMethod.Delete
    override val path: String = "/features/$featureId"
  }
}
